src/polymer/polymer-element.ts

```diff
@@ -334,52 +334,64 @@ function inheritFrom(element: PolymerElement, superElement: PolymerExtension) {
   // TODO(justinfagnani): fixup and use inheritValues, but it has slightly odd
   // semantics currently
 
-  for (const superProperty of superElement.properties) {
-    const newProperty = Object.assign({}, superProperty);
-    if (!newProperty.inheritedFrom) {
-      const superName = getSuperName(superElement);
-      if (superName) {
-        newProperty.inheritedFrom = superName;
-      }
-    }
-    element.properties.push(newProperty);
-  }
+  const superName = getSuperName(superElement);
+  _inheritVals(
+      element.properties, superElement.properties, superName, element.warnings);
+  _inheritVals(
+      element.methods, superElement.methods, superName, element.warnings);
+  _inheritVals(
+      element.attributes, superElement.attributes, superName, element.warnings);
+  _inheritVals(
+      element.events, superElement.events, superName, element.warnings);
 
-  for (const superMethod of superElement.methods) {
-    const newMethod = Object.assign({}, superMethod);
-    if (!newMethod.inheritedFrom) {
-      const superName = getSuperName(superElement);
-      if (superName) {
-        newMethod.inheritedFrom = superName;
-      }
-    }
-    element.methods.push(newMethod);
-  }
+  // TODO(justinfagnani): slots, listeners, observers, dom-module?
+  // What actually inherits?
+}
 
-  for (const superAttribute of superElement.attributes) {
-    const newAttribute = Object.assign({}, superAttribute);
-    if (!newAttribute.inheritedFrom) {
-      const superName = getSuperName(superElement);
-      if (superName) {
-        newAttribute.inheritedFrom = superName;
+type PropLike = {
+  name: string,
+  sourceRange?: SourceRange,
+  inheritedFrom?: string,
+  privacy?: Privacy
+};
+/**
+ * Mutates `existing`
+ *
+ * TODO(rictic): unify this with inheritValues in this file.
+ */
+function _inheritVals<P extends PropLike>(
+    existing: P[],
+    superVals: P[],
+    superName: string | undefined,
+    warnings: Warning[]) {
+  const existingVals = new Map(existing.map((e) => [e.name, e] as [string, P]));
+  for (const superVal of superVals) {
+    if (existingVals.has(superVal.name)) {
+      if (superVal.privacy === 'private') {
```

Comment:
Java is different than Javascript though because it does name resolution at compile time. When you declare a new field with the same name as a private field in Java they're totally separate. In Javascript code in the superclass that was expecting its private field will suddenly start getting the subclass's field, which if done on accident is a recipe for super weird bugs, pun intended.